src/interp.rs

```diff
@@ -19,342 +19,101 @@
+            filter: [[f32; $factor]; (TAPS / $factor)],
+            buffer: [F; (TAPS / $factor)],
+            buffer_pos: usize,
         }
 
-        fn reset(&mut self) {
-            // TODO: Use slice::fill() once stabilized
-            for v in &mut *self.z {
-                *v = 0.0;
+        impl<F> Default for $name<F>
+        where
+            F: FrameAccumulator + Default,
+        {
+            fn default() -> Self {
+                Self::new()
             }
-            self.zi = 0;
         }
 
-        fn process(&mut self, src: &[f32], dst: &mut [f32]) {
-            assert!(src.len().checked_mul(self.factor) == Some(dst.len()));
-            assert!(self.z.len() == self.delay * self.channels as usize);
-            assert!(self.filter.len() == self.factor);
-            assert!(self.zi < self.delay);
-
-            if src.is_empty() {
-                return;
-            }
-
-            let frames = src.len() / self.channels as usize;
-            let mut zi = self.zi;
-
-            for (src, (dst, z)) in src.chunks_exact(frames).zip(
-                dst.chunks_exact_mut(frames * self.factor)
-                    .zip(self.z.chunks_exact_mut(self.delay)),
-            ) {
-                zi = self.zi;
-
-                for (src, dst) in src.iter().zip(dst.chunks_exact_mut(self.factor)) {
-                    // Add sample to delay buffer
-                    //
-                    // TODO Ringbuffer without bounds checks for z/zi
-                    //
-                    // Safety: zi is checked to be between 0 and self.delay
-                    *unsafe { z.get_unchecked_mut(zi) } = *src;
-
-                    // Apply coefficients
-                    for (filter, dst) in self.filter.iter().zip(dst.iter_mut()) {
-                        let mut acc = 0.0;
-                        for (c, index) in &*filter.coeff {
-                            let i = (zi as i32 + *index as i32) % self.delay as i32;
-
-                            // Safety: zi is checked to be between 0 and self.delay
-                            acc += *unsafe { z.get_unchecked(i as usize) } as f64 * c;
-                        }
-
-                        *dst = acc as f32;
-                    }
-
-                    zi = (zi + self.delay - 1) % self.delay;
+        impl<F> $name<F>
+        where
+            F: FrameAccumulator + Default,
+        {
+            pub fn new() -> Self {
+                let mut filter: [[_; $factor]; (TAPS / $factor)] = Default::default();
+                for (j, coeff) in filter
+                    .iter_mut()
+                    .map(|x| x.iter_mut())
+                    .flatten()
+                    .enumerate()
+                {
+                    let j = j as f64;
+                    // Calculate Hanning window, with one tap ignored. (Last tap is zero anyways, and we want to hit
+                    // an even multiple of 48)
+                    let window = (TAPS - 1 + 1) as f64;
```

Comment:
`- 1 + 1` is strange :)

Reply:
I wrestled with it for a bit, and ended up writing it like that, to highlight the fact that we discard one zero-tap. I've changed it slightly, let's see if you find that less strange :)

Reply:
Better indeed